File: FSharp/HelloWorld/Lib/Extra/Rpn.fs
module HelloWorld.Lib.Extra.Rpn

open System

let private read s =
    Double.Parse s
    
let private inner_rpn stack op =
    match (op, stack) with
    | "+", n1 :: n2 :: tl -> n2 + n1 :: tl
    | "-", n1 :: n2 :: tl -> n2 - n1 :: tl
    | "*", n1 :: n2 :: tl -> n2 * n1 :: tl
    | "/", n1 :: n2 :: tl -> n2 / n1 :: tl
    | "^", n1 :: n2 :: tl -> Math.Pow(n2, n1) :: tl
    | "ln", n1 :: tl -> Math.Log(n1) :: tl
    | "log10", n1 :: tl -> Math.Log10(n1) :: tl
    | "sum", l -> [List.sum l]
    | "prod", l -> [List.fold (*) 1.0 l]
    | x, l -> read x :: l

let rpn (s: string) =
    try
        let res :: tl = List.fold inner_rpn [] (List.ofArray (s.Split(" ")))
        if (tl <> []) then
            failwith "Incorrect input"
        else
            res
    with
        | :? MatchFailureException as ex ->
            Console.Error.WriteLine ex.Message
            0.0

File: Haskell/HelloWorld/src/Extra/Rpn.hs
module Extra.Rpn where

import Prelude

import Data.Foldable ( Foldable(foldl') )
import GHC.Float ( powerFloat )

rpn :: String -> Either String Float
rpn input =
  let res : xs = foldl' (flip innerRpn) [] $ words input in
  case xs of
    [] -> Right res
    _ -> Left "Bad Input"

innerRpn :: String -> [Float] -> [Float]
innerRpn "+" (x : y : xs) = (y + x) : xs
innerRpn "-" (x : y : xs) = (y - x) : xs
innerRpn "*" (x : y : xs) = (y * x) : xs
innerRpn "/" (x : y : xs) = (y / x) : xs
innerRpn "^" (x : y : xs) = powerFloat y x : xs
innerRpn "ln" (x : xs) = log x : xs
innerRpn "log10" (x : xs) = logBase 10 x : xs
innerRpn "sum" xs = [sum xs]
innerRpn "prod" xs = [product xs]
innerRpn n xs = readFloat n : xs

readFloat :: String -> Float
readFloat s = read s :: Float